### ISSUE TITLE
Use Gradle to build a umple compiler jar

### ISSUE THREAD

**Issue by AdamBJ** [umple/umple.gradle#10]
At a high level, we need Gradle to:

1. "Compile" the `master.ump` files in the sub projects used by the plugin to generate the Java source the Umple compiler needs
2. Compile the generated Java source files and package them into a jar

See the discussion in https://github.com/umple/umple/issues/751#issuecomment-282145135 for details

**Comment by AdamBJ** [comment]
@Nava2 @vahdat-ab Let's use this issue to track the progress of the Umple compiler jar task

**Comment by AdamBJ** [comment]
Hi @Nava2 @vahdat-ab,

I'm trying to decide which repo I should put the compiler jar work in. I see two options.

1. Copy the directories I'll need into `umple.gradle` and see if I can get the compiler built in an "artificial" environment. I'd create a branch in `umple.gradle` if I did it this way.
2. Add `build.gradle` files to `cruise.umple` and try to build the compiler in the "real" environment. I'd create a branch in `cruise.umple` if I did it this way. 

What do you guys think? I think it makes sense to do it the second way. I'm trying to figure out how to setup the directory structure right now. Usually the root project contains all the subprojects, but in the case of Umple I don't think this will be the case. Do either of you know who to deal with a multi project build where the subprojects are scattered in different locations?

**Comment by AdamBJ** [comment]
Found something that looks promising: https://discuss.gradle.org/t/how-to-include-sub-projects-in-settings-gradle-file/5096. I'm exploring the solution they suggest now.

**Comment by AdamBJ** [comment]
I've got it working for `UmpleParser`. `UmpleParser`'s `master.ump` is being processed by the plugin, and the generated files are being written to `umple\cruise.umple\src-gen-umple`. Now I'm working on figuring out a couple pathing issues to make it easier to configure the root projects `settings.gradle` file. Good progress though!

**Comment by vahdat-ab** [comment]
I would say keep the main project inside the root folder called `gradle.build`.
subprojects should be landed in their own  projects.
Indeed, those subprojects should have tasks that allow other people to build that specific project alone. The reason for this is this:
UmplePraser was a part of Umple compiler and last year we made it independent. It means anyone can grab it and use it for different purposes. Therefore, that build file should allow other users to build it through gradle and integrate it to their own build system. (consider this functionality as a low priority task)

**Comment by Nava2** [comment]
I agree with @vahdat-ab on putting things in a `gradle.build` folder or similar. Once we can get the full build proven to replace the ant build, we can rename the folder and just have a `build` folder. That's a long way out, though! 👍

**Comment by AdamBJ** [comment]
Ok, good. We're all agreed on a general structure then. 

The problem I'm trying to overcome right now is dependencies. I've setup `UmpleParser` and `UmpleToJava` as sub projects of `gradle.build` in the way we agreed upon and have everything generating properly. The generated files are being outputted in the correct place in `cruise.umple` too, so that's good. However, When I try to compile I quickly enter dependency hell.

This is what the inside of `cruise.umple` looks like right now:

```
cruise.umple
     bin
     src-gen-umple
     src-gen-umpletl
     gradleTest
          bin
          src-gen-umple
          src-gen-umpletl
```

The folders in `gradleTestf` are copies of the originals in `cruise.umple`. The `UmpleParser` and `UmpleToJava` sub projects output their generated java files in the appropriate places in the copies of `src-gen-umple` and `src-gen-umpletl` inside `gradleTest`. I did it this way initially so that I could compile the java files from `UmpleToJava` and `UmpleParser` without having to worry about dependencies from other sub projects I haven't setup yet. 

This actually works for `UmpleParser`. It compiles nicely and produces all the `.class` files you'd expect. The problem is `UmpleToJava`. When it's compiled the compiler doesn't seem to know where to look for the files it needs. For example, it can't find the `cruise.umple.util` package. I suspect this is something to do with the way `generatedSource` source set's classpath. We could add everything we need for `UmpleToJava` to the `generatedSource` set's class path I guess, but doing that for multiple projects would be a nightmare.

I hope that it's not too complicated to have multiple compilation tasks like this. It does seem easier to just compile everything in a single task,, but let's keep that as a last resort for now. It would be nice to get the parallelization bump from multiple compilation tasks. If you guys have any ideas on how to solve the "one-at-a-time compilation" mess I'm all ears.

**Comment by Nava2** [comment]
How are you setting your compile dependencies?

**Comment by AdamBJ** [comment]
In `UmpleToJava`s `build.gradle` file:

```
dependencies {
    compile ...
}
```

Is what I've been experimenting with (replaced `...` with various options). I tried using the umple jar as a compile dependency, but that didn't work. I'm not sure what I should be listing exactly. Some kind of binary I guess, but I'm not sure what.

**Comment by AdamBJ** [comment]
I did it! Jeez, took me all day but I finally cracked it after reviewing some classpath stuff. I added the `bin` folder to `generatedSource`'s `compileClasspath` like so:

`sourceSets.generatedSource.compileClasspath = project.files('../cruise.umple/gradleTest/bin')`

Go figure, binary files are kept in the `bin` folder. That makes sense now, haha

**Comment by Nava2** [comment]
I think the idea is okay, there's definitely a property in Gradle to denote the build directory name. Probably should use that rather than bin by default.

**Comment by Nava2** [comment]
Well done on all of this. You're doing great work!

**Comment by AdamBJ** [comment]
Hi guys,

I've been thinking about how to get the most parallelization out of Gradle. I notice there's a `-parallel` flag we can use when building that [seems to build subproject in parallel](https://docs.gradle.org/current/userguide/multi_project_builds.html#sec:parallel_execution). However, it requires that all the project are "[decoupled](https://docs.gradle.org/current/userguide/multi_project_builds.html#sec:decoupled_projects)". Unfortunately, it seems we can't use the `subproject`s closure if we want the build project to be decoupled. That mean's we're going to need to reproduce a lot of boiler plate (e.g. `apply umple plugin`) across subprojects.

@Nava2, did you have using the `parallel` option in mind when you talked about doing parallel compilation of the java source? I wonder if there's a way to parallelize things while avoiding the boilerplate.

On an unrelated note, I'm doing the Umple compiler jar work in a branch of the main Umple project right now. Since the `build.gradle` files of the subprojects are embedded in the subproject folders of the main umple project I thought it made the most sense to do it that way. I haven't pushed anything yet, I'll let you guys know when I do.

**Comment by Nava2** [comment]
I feel its best to let gradle decide when to use parallel. If folks want to turn on the flag, let them. Ideally, we want things decoupled, but we can't always do that.

**Comment by AdamBJ** [comment]
I've pushed a prototype that shows (roughly) how I'm going to structure the umple jar project: https://github.com/AdamBJ/umple/commit/d2ba1a35e36dfa68312a48351000ed1c75e856ae

I wouldn't trying running it locally on your machine at this point -- there's a lot of duct-tape holding it together and I'm not sure how it would work outside of my local machine. One thing I'll do is change it so that the main `build.umple` file is in it's own folder (right now it's in the ant build folder).

The current roadblock I'm trying to overcome seems to be caused by a bug in Umple. `umple\cruise.umple\src-gen-umple\cruise\umple\docs\DocumenterMain.java` is getting an extra `}` added on at line 73. Because of the brace mismatch I'm unable to compile that file. It seems like the copy of `DocumenterMain.java` produced by a normal ant build also has this problem. However, when I run the corresponding Umple file (`Documenter_Code.ump`) on UmpleOnline, the extra brace disappears. It's driving me crazy! Do you guys have any ideas?

`DocumenterMain.java` is a weird file because its contents appear to be related to the context in which it was generated. For example, the `DocumenterMain.java` file generated by UmpleOnline is less than 1000 lines, whereas the version generated as part of a normal build is more than 2000 lines.

**Comment by AdamBJ** [comment]
Hey @vahdat-ab @Nava2,

I've been looking at how Ant does the umple jar compilation. Here's the important bit:

```
target name="compile">
    <echo message="${ant.project.name}: ${ant.file}"/>

    <deps-load-path conf="core"   pathid="core.ivy.classpath" />
    <deps-load-path conf="test"   pathid="test.ivy.classpath" />

    <javac debug="true" includeantruntime="false" debuglevel="source,lines,vars" destdir="${bin.path}" source="1.8" target="1.8">
	    <src path="${xtext.project.path}/src"/>
      <src path="${xtext.project.path}/src-gen"/>
      <src path="${project.path}/src"/>
      <src path="${project.path}/src-gen-umpletl"/>
      <src path="${project.path}/src-gen-umple"/>
      <src path="${project.path}/test"/>
      <src path="${vendors.path}/jopt-simple/src"/>
      <exclude name="**/.git"/>
      <exclude name="**/*.ump" />
      <exclude name="**/data" />
      <classpath refid="project.classpath"/>
      <classpath refid="validator.project.classpath"/>
      <classpath refid="core.ivy.classpath" />
      <classpath refid="test.ivy.classpath" />
	    <!-- Add compiler arguments here, see https://ant.apache.org/manual/using.html#arg for details, example below
			  <compilerarg value="-Xlint:deprecation" />
		-->
    </javac>
    
    <copy todir="${bin.path}" overwrite="true">
	    <fileset dir="${project.path}/src"><include name="**/*.grammar"/></fileset>
	    <fileset dir="${project.path}/src"><include name="**/*.error"/></fileset>
    </copy>
    <delete file="cruise.umple/src/rules.grammar"/>
    <delete file="cruise.umple/bin/rules.grammar"/>
  </target>
```

Basically, it specifies a bunch of src directories where the java files are stored and then compiles them all. It also does some class path manipulation that I'm working on emulating with Gradle. One thing I'm not sure of though is if there's an order that the src directories are compiled in. This is important for us - since we're compiling each sub project individually, if there are dependencies between projects we'll need to be aware of them when determining the order we want to compile sub projects in. Do either of you guys know if the src directories are compiled in any particular order? Or can they be compiled in any order as long as the ivy dependencies are present?

**Comment by AdamBJ** [comment]
Small update on the mysterious `}`. It seems that when I run `:CruiseUmple:compileGeneratedSourceJava` not only are the `cruise.umple` .java files within the `GradleTest` directory replaced, the `cruise.umple` .java files in the main `umple.gradle` directory are, too. I can't figure out why this is happening since I'm expliticly specifying in the `cruise.umple` build file that the output should be stored within `GradleTest`:

`project.ext.set("outputPath", "GradleTest/src-gen-umple")`

I think this behaviour may be the problem since when I do a full build `DocumenterMain.java` is correct. It's only once I've run the `CruiseUmple:compileGeneratedSourceJava` task a couple times that the extra `}` occurs.

Any ideas why it's overwriting the files in two places or what the heck is happening?

**Comment by Nava2** [comment]
> Basically, it specifies a bunch of src directories where the java files are stored and then compiles them all. It also does some class path manipulation that I'm working on emulating with Gradle. One thing I'm not sure of though is if there's an order that the src directories are compiled in. This is important for us - since we're compiling each sub project individually, if there are dependencies between projects we'll need to be aware of them when determining the order we want to compile sub projects in. Do either of you guys know if the src directories are compiled in any particular order? 

Normally, when working with Java projects we don't care about the order. In this case, you're right and we do have to care, now. I haven't the faintest what order of projects would be. Maybe @vahdat-ab knows. Though, I do believe it is more likely that you'll need to "guess and check" your way through that dependency graph. 

> Or can they be compiled in any order as long as the ivy dependencies are present?

Ivy just fills in the gaps for external dependencies, which you've already tackled some of them it looks like. You should be able to copy the ivy dependencies directly over. 

> Small update on the mysterious }. It seems that when I run :CruiseUmple:compileGeneratedSourceJava not only are the cruise.umple .java files within the GradleTest directory replaced, the cruise.umple .java files in the main umple.gradle directory are, too. I can't figure out why this is happening since I'm expliticly specifying in the cruise.umple build file that the output should be stored within GradleTest:
>
> I think this behaviour may be the problem since when I do a full build DocumenterMain.java is correct. It's only once I've run the CruiseUmple:compileGeneratedSourceJava task a couple times that the extra } occurs.
>
> Any ideas why it's overwriting the files in two places or what the heck is happening?

It most likely seems that you found a bug in Umple. If I had to guess about the multiple unmentioned files, there may be a hard-coded path somewhere that is forcing it to always write to a location. For the extra `}` the fact that it isn't deterministic is concerning.. could it be version differences between the umple compilers? that is, the first run uses a "more stable" version than the dev. version?

**Comment by Nava2** [comment]
I'm not a huge fan of forcing gradle to do strange things because of the project structure. Could you put together a proposal of what changes would need to be made to the project to make it more "gradle friendly?" These would include directory structure changes for the most part. 

Additionally, could we revisit the `project.ext.*` methods, looking to get those back into a configuration object ideally.

**Comment by vahdat-ab** [comment]
> However, when I run the corresponding Umple file (Documenter_Code.ump) on UmpleOnline, the extra brace disappears. It's driving me crazy! Do you guys have any ideas?

I have such issue with Eclipse plugin when I run two times a file which has a `main` method. Don't know why it happens


> Do either of you guys know if the src directories are compiled in any particular order? Or can they be compiled in any order as long as the ivy dependencies are present?

If you do compilation separately for each one you should be able to get it easily. According to my knowledge:
xtext.project.path is not supported anymore. We need such a thing for a new Eclipse plugin which is not `based on Xtext` The building of plugin must be done after the `jar` file is ready.
`{vendors.path}/jopt-simple/src` compiler and somehow UmpleTL depends on this (not sure exactly where).
Others must be explored. Techincally code generators must be independent but I think they are not based on what you told me before.
Another way is to just copy each project to a different folder and then use CLI to generate code and compile them. If that happens it means you can compile it separately through gradle.

NOTE: I personally wouldn't mind in this phase having those tasks depend on each other. These kinds of problems can be fixed late (by refactoring those projects). What important currently is to be able to have a working build project. Some of the issues you're getting come from the dimension that you don't have a control on them or they are time-consuming to being fixed in this period of time. 

> I think this behaviour may be the problem since when I do a full build DocumenterMain.java is correct. It's only once I've run the CruiseUmple:compileGeneratedSourceJava task a couple times that the extra } occurs.

My issue with Eclipse behaves like that. It works ok at the first time and then adds extra `{`


> For the extra } the fact that it isn't deterministic is concerning.. could it be version differences between the umple compilers? that is, the first run uses a "more stable" version than the dev. version?

I believe it happens because of a `static` variable in some part of the project or sth like that. I'm busy with my thesis otherwise, will jump into it to fix it  :-)

**Comment by AdamBJ** [comment]
@Nava2 

> I'm not a huge fan of forcing gradle to do strange things because of the project structure. Could you put together a proposal of what changes would need to be made to the project to make it more "gradle friendly?" These would include directory structure changes for the most part.

Here's the structure of the umple project:
```

umple
   cruise.umple
   ...
   UmpleToJava
   ...
   UmpleParser
   build.gradle //root project's build file
   settings.gradle //root project's settings file
```

As long as I put the `build.gradle` and `settings.gradle` file for the root project inside the main `umple` folder we don't need to overwrite the paths for many if not all of the sub projects. Putting the root project's `build.gradle` file directly inside `umple` makes Gradle think that `umple` is the root project directory. Since the sub project directories are directly beneath the `umple` directory, this works.

The only case where it might not work is if we have a sub project that isn't directly below the `umple` folder. I'm not sure if we have any sub projects of this type. 

One reason to keep the `switch` is because it makes the sub project build files much easier to work with. I often have 3 or 4 build files open in my editor at once, and if they're all called `build.gradle` it makes things awkward. Are you ok with renaming the sub project build files?

> Additionally, could we revisit the project.ext.* methods, looking to get those back into a configuration object ideally.

I found a way to attach the settings to the generatedSource source set: https://github.com/umple/umple.gradle/issues/8. It uses the exact same `ext` mechanism, but at least it stores the information closer to where it's used. What do you think of that?

**Comment by AdamBJ** [comment]
The issue with the extra `}` is very frustrating and will be a blocker for us not too far down the road. For now, I'm going to try and leave the `cruise.umple` sub project until the end. Let's just hope that none of the other sub projects have the same kind of problem.

**Comment by vahdat-ab** [comment]
@AdamBJ NOTE: the issue with `{` doesn't show up if I run Umple.jar through the command line. It works perfectly.   It happens when one of the main classes is called through the code. The plugin doesn't call jar and it just uses the compiled classes.

**Comment by AdamBJ** [comment]
@vahdat-ab Doesn't the umple jar contain the compiled classes? I thought that the plugin gets the `.class` files it needs from the umple jar. 

Are you suggesting making the plugin somehow use the jar from the command line?

**Comment by vahdat-ab** [comment]
@AdamBJ I don't suggest because in that case, we will get less flexibility in the plugin later. we need to find the reason. The compiler works correctly when it's packed in a jar.  When we use classes it happens (this is just based on my initial investigation).

**Comment by AdamBJ** [comment]
It seems like all the `UmpleTo___` projects depend on `cruise.umple` so we're going to have to find a fix for the `}` issue. It seems like rebuilding the jar from the command line fixes things temporarily, so I'll use that as a workaround for now. But it's a couple minutes to wait for everything to build and generate each time I want to test a change 😢

**Comment by Nava2** [comment]
So could it be because the plugin is calling the Umple calls directly, aka something wierd is happening in umple's "cli" loop? I'm spit balling here. 

For the project structure, 

> As long as I put the build.gradle and settings.gradle file for the root project inside the main umple folder we don't need to overwrite the paths for many if not all of the sub projects. Putting the root project's build.gradle file directly inside umple makes Gradle think that umple is the root project directory. Since the sub project directories are directly beneath the umple directory, this works.

I would rather have them all called `build.gradle` to follow convention. It keeps the `settings.gradle` simpler, but I could see arguments both ways. _shrug_. 

> It seems like all the UmpleTo___ projects depend on cruise.umple so we're going to have to find a fix for the } issue. It seems like rebuilding the jar from the command line fixes things temporarily, so I'll use that as a workaround for now. But it's a couple minutes to wait for everything to build and generate each time I want to test a change 😢

Nice thing is that gradle does do incremental builds, where ant doesn't. 😅

**Comment by vahdat-ab** [comment]
>  It seems like rebuilding the jar from the command line fixes things temporarily,

That's the beauty of plugins. Go with that until you figure out the issue. I don't want you to spend your time on that issue. 

> So could it be because the plugin is calling the Umple calls directly, aka something wierd is happening in umple's "cli" loop? I'm spit balling here.

I don't think so because in the Eclipse plugin there is no loop and also I create always a new model and then run it.

@AdamBJ I wonder if the file you pointed out has a `main` method? In my case, it just happens when there is a `main` function in the class.

**Comment by AdamBJ** [comment]
@vahdat-ab Yes, it has a main method. I guess it must be the same issue you're experiencing. I'm trying to get incremental builds working right now so I can avoid having to wait for the jar to rebuild every time I want to make a change.

**Comment by vahdat-ab** [comment]
@AdamBJ then ok. I talked to you about the possible source of the issue on skype. Check that one.

**Comment by vahdat-ab** [comment]
@AdamBJ In the test case [DistributedClassTest](https://github.com/umple/umple/blob/e17650ee8379f5b0897139d20337ba11c7933fce/cruise.umple/test/cruise/umple/implementation/DistributedClassTest.java).java, you added the following line:
```
public void tearDown() {
	/* Nullify mainMainClass. It's a static variable, if we don't do this the state will
	 * affect the next set of JUnit tests that use mainMainClass.
	 */

	JavaClassGenerator.mainMainClass = null;
  }	`
```
The issue might be because of `mainMainClass `

**Comment by AdamBJ** [comment]
Hey guys!

I made some good progress on getting the ivy dependencies transferred over to the Gradle build. What I've done is added the `core` ivy dependencies (ant and jopt-simple) to the generatedSource source set's `compileClasspath`:

The only way I could figure out how to do that was to make a `configuration` and then add the configuration to the source set's `compileClasspath` like this (in `CruiseUmple.build`):

```
configurations {
    coredeps
}

dependencies {
    coredeps group: 'net.sf.jopt-simple', name: 'jopt-simple', version: '4.4'
    coredeps group: 'org.apache.ant', name: 'ant', version: '1.9.6'
    coredeps group: 'junit', name: 'junit', version: '4.12'
} 

sourceSets.generatedSource.compileClasspath += configurations.coredeps
```

I originally was trying this:

```
dependencies {
    generatedSourceCompile group: 'net.sf.jopt-simple', name: 'jopt-simple', version: '4.4'
    generatedSourceCompile group: 'org.apache.ant', name: 'ant', version: '1.9.6'
    generatedSourceCompile group: 'junit', name: 'junit', version: '4.12'
} 
```
But for some reason that was only adding the deps to the run time classpath.

Anyways, the end result is that `cruise.umple` is now generating and building properly! The only thing I'm not clear on is how to deal with the `cruise.umple/bin` compile classpath dependency for `cruise.umple`. The bin file is actually generated by compilation, so by adding `bin` as a compile classpath dependency I think I would be using the `bin` from the last build. This seems wrong to me, but the ant build seems to be doing it this way so I'm not sure. What do you guys think? The alternative is to try and compile whatever sub projects `cruise.umple` depends on before trying to compile `cruise.umple`. This would involve a lot of guess-and-check though to figure out what projects depend on each other.

**Comment by AdamBJ** [comment]
@Nava2 @vahdat-ab Here's a prototype build file for `UmpleToJava` that shows some of the techniques I'm using to get things built:

```
// Configure the plugin
apply plugin: 'java'
sourceSets {
    generatedSource { 
       ext.languageToGenerate = 'Java' 
       ext.outputPath = '../cruise.umple/GradleTest/src-gen-umpletl/cruise/umple/compiler/java'
       ext.umpleFilePath = 'UmpleTLTemplates/Master.ump'
    }
}

// add compile classpath dependency for the generated source source set (ivy dependencies can also be added this way)
sourceSets.generatedSource.compileClasspath = project.files('../cruise.umple/bin') 
// after we compile the `generatedSource` SS we output output .class files here
sourceSets.generatedSource.output.classesDir = "../cruise.umple/GradleTest/bin" 
// workaround that allows for incremental builds. If a SS doesn't have a srcDir then the compile task is always skipped
sourceSets.generatedSource.java.srcDir "../cruise.umple/GradleTest/src-gen-umpletl/cruise/umple/compiler/java" 

// More incremental build stuff
generateSource {
   // generateSource task only runs if the contents of this directory or `srcDir` have changed
    outputs.dir("../cruise.umple/GradleTest/src-gen-umpletl/cruise/umple/compiler/java") 
}

```

This is admittedly rough. Things could probably be grouped much more nicely using closures, but this post and my previous post shows two things:

1) We can do incremental builds (I haven't tested this fully, but it seems to be working)
2) We can add whatever dependencies required to compile the generated source files in a pretty easy way. 

Using the techniques outlined above, I've gotten `UmpleParser`, `cruise.umple`, and `UmpleToJava` generating, compiling, and building correctly. There are still some outstanding issues like the issue with `bin` I talked about in my previous post, and the syntax of the build files should be refined. But I think we now have the tool we need to get the Umple compiler jar built.

**Comment by AdamBJ** [comment]
Quick update. I'm still am not sure the reason `cruise.umple/bin` seems to be on the compile classpath. I've ask Tim about it and he didn't seem to know either. I've decided to ignore it for now, as the idea of preserving `.class` files between compilation runs seems ludicrous.

I've made some progress in determining the dependencies between the projects. It seems like the `UmpleTo_` all depend on `cruise.umple`, and perhaps `UmpleParser`. I haven't verified this for all the TL projects, but it's true for `UmpleToJava`. When I compile `cruise.umple` before `UmpleToJava` it works. If I reverse the order, it doesn't. 

The big remaining question mark now is what `cruise.umple` depends on. It definitely depends on something, because I can't compile it unless I've got `cruise.umple/bin` on the compile classpath. My suspicion is that it somehow depends on the Umplificator or Validator projects, but I've yet to try building those two. I'll try building those next and see if my hunch is accurate.

**Comment by vahdat-ab** [comment]
> The big remaining question mark now is what cruise.umple depends on. 

Can't you get it from Ivy dependencies? One of them is `cruise.umple.nebula`

**Comment by AdamBJ** [comment]
@vahdat-ab where do you see that? I see we're using the `core` and `test` Ivy dependencies, but those two configurations make no mention of `cruise.umple.nebula`. If `cruise.umple.nebula` is a dependency of `cruise.umple` that's very good to know. The next question is what are the dependencies of `cruise.umple.nebula`! I can try to figure that out by trial and error, but if you're aware of a dependency file that I'm missing please let me know. It'll make my life easier.

**Comment by vahdat-ab** [comment]
@AdamBJ unfortunately, people who worked on those dependencies have never documented and I just know that from my previous experience. To me, the best person who can comment on this is @aforward

**Comment by AdamBJ** [comment]
Ok. @aforward, if you have any knowledge about the dependencies between the Umple sub projects I'd really appreciate hearing about it!

**Comment by AdamBJ** [comment]
This is getting messier and messier. It appears the `UmpleToJava` depends on at least of a portion of `cruise.umple`. `cruise.umple` produces `cruise.umple.util`, which `UmpleToJava` needs:

```
:UmpleToJava:compileGeneratedSourceJava
C:\Users\i_am_\workspace\umple\cruise.umple\GradleTest\src-gen-umpletl\cruise\umple\compiler\java\JavaClassGenerator.java:6: error: package cruise.umple.util does not exist
import cruise.umple.util.*;
```

However, it also appears that `cruise.umple` needs all of the `UmpleTo__` stuff!

Here's one of the error message I get if the `UmpleTo__` stuff isn't on the classpath:

```
package cruise.umple.compiler.java does not exist
import cruise.umple.compiler.java.*;
^
C:\Users\i_am_\workspace\umple\cruise.umple\GradleTest\src-gen-umple\cruise\umpl                                                                                                                                                                                      e\compiler\PhpGenerator.java:9: error: package cruise.umple.compiler.php does no                                                                                                                                                                                      t exist

```

So `UmpleToJava` needs to be build before `cruise.umple` (or a least the portion `util` portion of `cruise.umple`), and `cruise.umple` needs to be built before `UmpleToJava`. What a mess 😭

**Comment by AdamBJ** [comment]
@vahdat-ab @Nava2 Guys, I'm not sure how well the "compile each sub project individually" plan is going to work. Like I mentioned in my last post here, there appears to be some circular dependencies between sub projects that make individual compilation very difficult. I've been looking at how the Ant build does it and it seems like Ant compiles *all* the sub projects at once. Since all the sub projects are aware of each other at the time of compilation, there aren't any dependency issues. As soon as we try to treat each sub project separately though, we run into trouble. 

I'm going to keep digging to see if there's any way around this problem. But in the worst case I think we'll have to do some, if not most, of the compilation in a single task a la Ant. That's a pretty huge change to be making this late in the game, though I think it's possible to get done before the semester is over. We already know we can generate all the `.java` files we need. It'd just be a matter of coming up with a "mega compile" task that compiles everything at once. What do you guys think?

**Comment by Nava2** [comment]
This screams like something we're better off doing right than quick. I think we need to figure out _why_ sub-projects depend on others and consider making a plan to start isolating those components better. This is outside my domain and reaching into @TimLethbridge and @vahdat-ab's. 

If we need to do some refactoring of the projects themselves to extract commonality or define "better" groupings, this might be the time to do it. However, that is *not* a small undertaking and I _highly_ doubt it's doable by the end of the semester.

**Comment by AdamBJ** [comment]
I agree that completely untangling the dependency web is most likely beyond the scope of this semester. Having dipped my toe in that task, I think it'll take a lot of patience and time to get it done right. I believe @TimLethbridge mentioned that the UmpleParser was pulled out last year and that it took some effort. If we need to do something similar for the other projects it could be a big job indeed.

One thing that might be more realistic is coming up with a dependency map. If we can figure out which sub projects depend on each other, that'll go a long way in determining the feasibility of actually separating out each of the projects so that they can be compiled individually.

**Comment by vahdat-ab** [comment]
My suggestion is this:
1) generate code for each project and if it is possible compiler it independently. This is gonna happen to some projects.
2) generate code for the rest and build the final system

This can make the project get the shape and you can move further. 

3) Highlight those projects and I'll assign them to someone in the future to refactor them. This is not sth that can be happened now along with this project. 
Once they are independent, we can modify the Gradle to get the benefit of what you have achieved so far.
I know we want to do everything in a perfect form, but sometimes some problems need to be resolved later as we progress. You have the infrastructure so no worries about the future modifications.

BTW, can you list me exactly which projects cannot be compiled independently?

**Comment by Nava2** [comment]
> One thing that might be more realistic is coming up with a dependency map. If we can figure out which sub projects depend on each other, that'll go a long way in determining the feasibility of actually separating out each of the projects so that they can be compiled individually.

I think that is an excellent idea. If we can figure out the classes in each project and map them to the project they live in, that'd be extremely useful in figuring out how to pull them apart. Easiest way is probably using a Java dependency analyzer, however, there may be something within Umple to figure that out, too (the graphviz stuff comes to mind). If it means augmenting something on top of Umple, it might be worthwhile. 

> My suggestion is this:
>
> 1. generate code for each project and if it is possible compiler it independently. This is gonna happen to some projects.
> 2. generate code for the rest and build the final system
> 
> This can make the project get the shape and you can move further.
>
> 3. Highlight those projects and I'll assign them to someone in the future to refactor them. This is not sth that can be happened now along with this project.
>    Once they are independent, we can modify the Gradle to get the benefit of what you have achieved so far.

I agree entirely. 

> I know we want to do everything in a perfect form, but sometimes some problems need to be resolved later as we progress. You have the infrastructure so no worries about the future modifications.

Is this a case were we _need_ to do it quickly though? I agree to an extent that sometimes we want to push through to get it working, but in this context, we have a currently working system (ant), perhaps we can hold off? I'm okay with either option, but just playing devil's advocate.

**Comment by vahdat-ab** [comment]
>  I'm okay with either option, but just playing devil's advocate.

My point is this: since @AdamBJ has implemented the mechanism to generate and compile sub projects independently, so we have the infrastructure for the future improvement without worries regarding if it is doable or not. It is not like Ant that there is no way to go. 
Once the dependencies are resolved we get what you guys have been trying hard to achieve. I know how hard resolving dependencies is in Umple so that's the reason I'm saying maybe we can partially move on and try to bring the project to the point that if someone resolves the dependencies, the Gradle build can work much much better.

**Comment by AdamBJ** [comment]
Ok, having taken everything you guys have said into account, here's what I'm going to:

1) Get the Umple compiler jar built. I'll independently compile as many of the projects as possible, but for those that I can't compile independently I'll do as a group.

2) Come up with a dependency map to help with the refactoring we'll do down the road.

There's obviously some overlap between these tasks, but my priority will be to work on 1) rather than come up with a super detailed dependency map. Once I've completed 1) I should at least have a fairly good idea of which sub projects depend on each other -- the class-level stuff can be fleshed out with whatever time remains after we've got the jar is building properly.

A few other small things. @vahdat-ab to build the jar we need all the `UmpleTo__` projects, the `Umple Parser`, `cruise.umple`, the validator, and the umplificator project. I know that `cruise.umple` and the `UmpleTo__` projects need to be built together and that the Parser can be built separately. So in the best case we'll have 4 compilation tasks (for the `UmpleParser`, the validator, the umplificator, and everything else). In the worst case we'll have 2 tasks (one for the `UmpleParser` and one for everything else).

@Nava2 If we're going to be compiling multiple sub projects at once, do you think it would be worth adding support for multiple `master.ump` files to the plugin? That way we could have, for example, a single project to generate and compile `cruise.umple` and all the `UmpleTo__` projects. The plugin would generate all the java files and add everything it generated to the `generatedSource` source set. I *think* that would then allow us to compile files for both project by running the `compileGeneratedSourceJava` task.

If we don't take this approach we'd need to manually add all the generated source files to a source set, then run the `compile<SourceSet name>Java` task. I'm in favour of the first approach though because I think that being able to feed multiple `master.ump` files to the plugin would be a useful feature.

Now just keep your fingers crossed that it's not too hard to tell Gradle to include another project in a compilation task 😅

**Comment by AdamBJ** [comment]
Also, what is the purpose of the validator and umplificator projects? They're handled in a different way than the other projects, I think if I knew what they do I could make more sense of the special treatment they're getting.

The good news is that I don't think they're required to do a quick build. One of the Ant targets generates the Java files for them, but they aren't actually compiled. I should be careful what I say here, but I think getting a "quick build" done using Gradle shouldn't be too bad. The trickiest part may be figuring out how to get Gradle to make a runnable jar for us. Haven't done anything like that yet.

**Comment by vahdat-ab** [comment]
> If we're going to be compiling multiple sub projects at once, do you think it would be worth adding support for multiple master.ump files to the plugin?

This is a basic feature of the Umple compiler. You can feed several `.ump` file and the compiler will generate the code fo all of them. This is gonna be super easy to do. This feature should also be available in the plugin. BTW, compiling several `ump` files was my first issue fixed when I joint to the cruise team :-)

Regarding  

>  the validator and umplificator projects?

I'm gonna send an email to the guy who worked on these. I'll ask for the help. I'll also cc you as well.

**Comment by Nava2** [comment]
> This is a basic feature of the Umple compiler. You can feed several .ump file and the compiler will generate the code fo all of them. This is gonna be super easy to do. This feature should also be available in the plugin. BTW, compiling several ump files was my first issue fixed when I joint to the cruise team :-)

As @vahdat-ab said, yeup. Definitely. Default it to a single `Master.ump`, but definitely it should be a List of values. 

> @Nava2 If we're going to be compiling multiple sub projects at once, do you think it would be worth adding support for multiple master.ump files to the plugin? That way we could have, for example, a single project to generate and compile cruise.umple and all the UmpleTo__ projects. The plugin would generate all the java files and add everything it generated to the generatedSource source set. I think that would then allow us to compile files for both project by running the compileGeneratedSourceJava task.
> 
> If we don't take this approach we'd need to manually add all the generated source files to a source set, then run the compile<SourceSet name>Java task. I'm in favour of the first approach though because I think that being able to feed multiple master.ump files to the plugin would be a useful feature.

Here's an example where I would rather "hack it" for the umple project itself. I don't like the idea of multiple `master.ump` files across projects. I would much rather we just have the code generated, then manually merge the source sets and compile them with Java. It is ugly and hacky but in theory, it fixes itself when we fix the inter-project dependency issue.

**Comment by AdamBJ** [comment]
Ok, we'll add the ability to support multiple `.ump` files to the plugin, but stick to the "one subproject, one buildfile" rule for the Umple build. I've created a issue for the plugin upgrade: https://github.com/umple/umple.gradle/issues/16

**Comment by AdamBJ** [comment]
@Nava2 @vahdat-ab Here's the rough quick-build: https://github.com/AdamBJ/umple/commits/Umple_Gradle_Build_PoC

It works, aside from the issues described in #17. The way I got around the code generation problems for testing purposes was by using the source generated by the normal Ant build. When the source has been generated properly the Gradle build compiles and packages everything up into a complete compiler jar that can be used normally. Until the generation stuff is resolved I don't recommend you guys try and get the build running on your machines. Just have a look at how the build works

Whether I'm able to help resolve the Umple generation issues or not, this build should be really helpful for whoever continues on with the Gradle work. I've opened another issue to track the work that needs to be done next #20

**Comment by Nava2** [comment]
* "customMasterPath"?
* There looks like a lot of common code that just changes by name, is that something we could use a variable for?
* [this](https://github.com/umple/umple/compare/master...AdamBJ:Umple_Gradle_Build_PoC#diff-c197962302397baf3a4cc36463dce5eaR58) could be done in each lower build script instead
* I'd rather [this](https://github.com/umple/umple/compare/master...AdamBJ:Umple_Gradle_Build_PoC#diff-c197962302397baf3a4cc36463dce5eaR64) was done via normal gradle dependencies
* [File name is wrong?](https://github.com/umple/umple/compare/master...AdamBJ:Umple_Gradle_Build_PoC#diff-ebb3b9e2a7d55c2b63abfa1af9e1f696R1)
* Why not use the standard `build.gradle`? [Link](https://github.com/umple/umple/compare/master...AdamBJ:Umple_Gradle_Build_PoC#diff-88b7c47e47b8ee65263b6784b86fa0a7R24)

**Comment by AdamBJ** [comment]
> "customMasterPath"?

I've got some plugin changes that I haven't pushed yet. That flag allows the user to override the `src/{source set name}/umple convention`. It's false by default and if someone chooses not to use it the plugin operates in the same way as always. 

I'll provide more details when I push, but the way the plugin was before, the user had no choice but to use the `src/{source set name}/umple convention`. They couldn't even use the `../../` workaround the way we had it setup.

>There looks like a lot of common code that just changes by name, is that something we could use a variable for?

Almost certainly. I was going for a PoC with this build.

> this could be done in each lower build script instead

I don't think it can, what do you mean?

>I'd rather this was done via normal gradle dependencies

That was the only way I could figure out how to get it working. Not sure how to do it the "normal" way

> File name is wrong?

Goes back to the `customMasterPath` flag. If that flag is set the user specifies a custom path relative to the project root.

> Why not use the standard build.gradle? Link

Just more convenient to edit/view multiple build files at once. Having seven `build.gradle`s open at once makes it hard to keep track of what project you're working on.
